### PR TITLE
SNOW-1524245 Gcm encryption using BouncyCastle

### PR DESCRIPTION
**Pull request by sfc-gh-knozderko**
### Description
SNOW-1524245 Gcm encryption using BouncyCastle

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../blob/master/CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [x] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+using Snowflake.Data.Core;
+using Snowflake.Data.Core.FileTransfer;
+using Snowflake.Data.Tests.Util;
+
+namespace Snowflake.Data.Tests.UnitTests
+{
+    [TestFixture]
+    public class GcmEncryptionProviderTest
+    {
+        private const string PlainText = "there is no rose without a smoke";
```
[the remark this answers is not in the document]
Comment (sfc-gh-dszmolka) [reply Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs:15]: or is a very elegant malaphor :)

File: Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        private static readonly byte[] s_plainTextBytes = Encoding.UTF8.GetBytes(PlainText);
+        private static readonly byte[] s_qsmkBytes = TestDataGenarator.NextBytes(GcmEncryptionProvider.BlockSizeInBytes);
+        private static readonly string s_qsmk = Convert.ToBase64String(s_qsmkBytes);
+        private static readonly string s_queryId = Guid.NewGuid().ToString();
+        private const long SmkId = 1234L;
+        private const string KeyAad = "key additional information";
```
[the remark this answers is not in the document]
Comment (sfc-gh-knozderko) [reply Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs:21]: we do have such a test: `TestEncryptAndDecryptWithoutAad`

File: Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+                {
+                    // assert
+                    var decryptedText = ExtractContent(decryptedStream);
+                    CollectionAssert.AreEqual(s_plainTextBytes, decryptedText);
+                }
+            }
+        }
+
+        [Test]
+        public void TestEncryptAndDEncryptWithAad()
```
[the remark this answers is not in the document]
Comment (sfc-gh-knozderko) [reply Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs:77]: done

File: Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs:21]: No, this is a test with nulls, what I mean is empty strings.

[the remark this answers is not in the document]
Comment (sfc-gh-knozderko) [reply Snowflake.Data.Tests/UnitTests/GcmEncryptionProviderTest.cs:21]: I added TestEncryptAndDecryptWithEmptyAad

**Review by sfc-gh-pfus — APPROVED** [review]

**Review by sfc-gh-dstempniak — APPROVED** [review]

**Pull request merged**